### PR TITLE
Anod actions implementation

### PR DESCRIPTION
**Pull request by piratos**

### CODE REVIEW

File: e3/electrolyt/run.py [hunk 1]
```diff
@@ -28,6 +34,90 @@ def run(self):
...
+                logger.debug('%s created in cache/tmp' % source_name)
+                return
+
+    def do_getsource(self):
+        """action_item from an intermediate node."""
+        self.status = SUCCESS
+
+    def do_installsource(self):
+        """Install the source from tmp/cache to build_space/src."""
+        spec = self.data.data[0]
```
Comment (enzbang) [inline e3/electrolyt/run.py]: can it be something else that a list?

File: e3/electrolyt/run.py [hunk 2]
```diff
@@ -28,6 +34,90 @@ def run(self):
...
+        spec.sandbox = self.sandbox
+        anod_instance = AnodDriver(spec, None)
```
Comment (enzbang) [inline e3/electrolyt/run.py]: see previous comment wrt AnodDriver

File: e3/electrolyt/run.py [hunk 3]
```diff
@@ -28,6 +34,90 @@ def run(self):
...
+        source = self.data.data[1]
+        src_dir = os.path.join(self.sandbox.tmp_dir,
+                               'cache',
+                               source.name)
+        dest_dir = os.path.join(spec.build_space.src_dir)
+        sync_tree(src_dir, dest_dir, ignore=source.ignore)
+        self.status = SUCCESS
+
+    def do_root(self):
+        """Epress the final result of the exec."""
```
Comment (enzbang) [inline e3/electrolyt/run.py]: ?

File: e3/electrolyt/run.py [hunk 4]
```diff
@@ -28,6 +34,90 @@ def run(self):
...
+        # If nothing fails in the process the status will remain automatically
+        # UKNOWN, as the root node UNKNOWN status is a SUCCESS
```
Comment (enzbang) [inline e3/electrolyt/run.py]: UNKNOWN

File: e3/electrolyt/run.py [hunk 5]
```diff
@@ -28,6 +34,90 @@ def run(self):
...
+        if self.status == STATUS_UNKNOWN:
```
Comment (enzbang) [inline e3/electrolyt/run.py]: I'd like to understand that. Why is it UNKNOWN?

File: tests/conftest.py [hunk 6]
```diff
@@ -30,7 +30,7 @@ def env_protect(request):
 
     def restore_env():
         Env().restore()
-        rm(tempd, True)
+#        rm(tempd, True)
```
Comment (enzbang) [inline tests/conftest.py]: hum... ;)

File: tests/tests_e3/anod/test_sandbox.py [hunk 7]
```diff
@@ -243,3 +244,53 @@ def test_anod_plan(git_specs_dir):
                'test e3 for %s' % platform]
     for action in actions:
         assert action in p.out
+
+
+def test_anodtest(git_specs_dir):
+    """Test the procedure of anodtest('e3')."""
```
Comment (enzbang) [inline tests/tests_e3/anod/test_sandbox.py]: use something simpler for our testsuite, we don't want to clone a remote repository

File: e3/anod/context.py [hunk 8]
```diff
@@ -434,6 +438,30 @@ def always_download_source_resolver(cls, action, decision):
                     (action.uid, decision.get_expected_decision())
             raise SchedulingError(msg)
 
+    @classmethod
+    def always_create_source_resolver(cls, action, decision):
+        """Froce source creation when scheduling a plan"""
```
Comment (Nikokrock) [inline e3/anod/context.py]: Type

File: e3/electrolyt/run.py (hunk 5 above)
Comment (piratos) [inline e3/electrolyt/run.py]: the status of the job is unknown by default if none of its predecessors has failed

File: e3/anod/context.py (hunk 8 above)
Comment (piratos) [inline e3/anod/context.py]: Done

File: e3/anod/sandbox/action.py [hunk 9]
```diff
@@ -195,6 +200,12 @@ def run(self, args):
 
             env = BaseEnv()
             cm = PlanContext(server=env)
+            resolver = AnodContext.always_create_source_resolver
+            if args.resolver:
+                if args.resolver in dir(AnodContext):
+                    resolver = getattr(AnodContext, args.resolver)
```
Comment (enzbang) [inline e3/anod/sandbox/action.py]: remove the if args.resolver in dir(...) and use:

resolver = getattr(AnodContext, args.resolver, AnodContext.always_create_source_resolver)

File: e3/anod/sandbox/action.py [hunk 10]
```diff
@@ -200,6 +204,9 @@ def run(self, args):
 
             env = BaseEnv()
             cm = PlanContext(server=env)
+            dry_run = False
+            if args.dry_run:
+                dry_run=True
```
Comment (enzbang) [inline e3/anod/sandbox/action.py]: why do you need that? what's wrong with args.dry_run directly?

File: e3/electrolyt/run.py [hunk 11]
```diff
@@ -15,26 +15,31 @@
 class ElectrolytJob(Job):
 
     def __init__(self, uid, data, notify_end, sandbox,
-                 force_status=STATUS_UNKNOWN):
+                 force_status=STATUS_UNKNOWN,
+                 dry_run=False):
```
Comment (enzbang) [inline e3/electrolyt/run.py]: this will require docstring + type hinting

File: e3/anod/action/__init__.py [hunk 12]
```diff
@@ -82,13 +82,15 @@ class InstallSource(Action):
     a GetSource Action.
     """
 
-    __slots__ = ('uid', 'data')
+    __slots__ = ('uid', 'spec', 'source')
 
-    def __init__(self, uid, data):
-        super(InstallSource, self).__init__(uid, data)
+    def __init__(self, uid, spec, source):
+        super(InstallSource, self).__init__(uid, data = (spec, source))
```
Comment (enzbang) [inline e3/anod/action/__init__.py]: no spaces around =

Comment (enzbang) [inline e3/anod/action/__init__.py]: And why do you keep data here?

File: e3/anod/status.py [hunk 13]
```diff
@@ -10,3 +10,6 @@ class ReturnValue(Enum):
     failure = 1
     missing = 2
     notready = 75
+    FORCE_SKIP = -125
```
Comment (enzbang) [inline e3/anod/status.py]: lowercase here

File: tests/tests_e3/anod/action/main_test.py [hunk 14]
```diff
@@ -13,6 +13,9 @@ def test_initall():
 
     class MySource():
         name = 'my_source'
+    class Spec(e3.anod.spec.Anod):
```
Comment (enzbang) [inline tests/tests_e3/anod/action/main_test.py]: missing blank line between classes?

File: tests/tests_e3/anod/action/main_test.py [hunk 15]
```diff
@@ -23,18 +26,16 @@ class MySource():
...
-    class Spec(e3.anod.spec.Anod):
-        uid = 'my_source_uid'
-        name = 'my_source_spec'
-
     create_source = action.CreateSource(spec=Spec(qualifier='', kind='source'),
                                         source_name='my_source')
     assert str(create_source) == 'create source my_source'
 
-    checkout = action.Checkout(repository='e3-core')
+    checkout = action.Checkout(repository=['e3-core'])
```
Comment (enzbang) [inline tests/tests_e3/anod/action/main_test.py]: why is it a list?

File: tests/tests_e3/anod/context_test.py [hunk 16]
```diff
@@ -77,13 +80,12 @@ def test_add_anod_action2(self):
 
         result = ac.schedule(ac.always_download_source_resolver)
         assert len(result) == 5, result.as_dot()
-
+        print(set(result.vertex_data.keys()))
```
Comment (enzbang) [inline tests/tests_e3/anod/context_test.py]: why?

File: e3/anod/context.py [hunk 17]
```diff
@@ -335,6 +335,8 @@ def add_spec(self,
                                                  e.kind,
                                                  e.qualifier)
 
+                    spec.deps[e.local_name] = result.data
```
Comment (enzbang) [inline e3/anod/context.py]: I don't think that data is a good name here. Is it possible to change it?

File: e3/electrolyt/run.py [hunk 18]
```diff
@@ -29,6 +34,97 @@ def run(self):
                 if hasattr(self, method_name):
                     getattr(self, method_name)()
 
+    def do_build(self):
+        """Run the build primitive after setting up the sandbox."""
+        self.data.data.sandbox = self.sandbox
```
Comment (enzbang) [inline e3/electrolyt/run.py]: and these data.data would be hard to debug!

File: e3/electrolyt/run.py [hunk 19]
```diff
@@ -29,6 +34,97 @@ def run(self):
...
+        anod_driver = AnodDriver(anod_instance=self.data.data, store=self.store)
+        anod_driver.activate()
+        anod_driver.anod_instance.build_space.create(quiet=True)
+        if not getattr(anod_driver.anod_instance, 'build', False):
```
Comment (enzbang) [inline e3/electrolyt/run.py]: Very minor but I would write it as:

 if getattr(anod_driver.anod_instance, 'build', None) is not None

File: e3/electrolyt/run.py [hunk 20]
```diff
@@ -29,6 +34,97 @@ def run(self):
...
+        anod_driver.anod_instance.build()
+        self.status = STATUS.success
+
+    def do_test(self):
+        """Run the test primitive."""
+        self.data.data.sandbox = self.sandbox
+        anod_driver = AnodDriver(anod_instance=self.data.data, store=self.store)
+        anod_driver.activate()
+        anod_driver.anod_instance.build_space.create(quiet=True)
+        if not getattr(anod_driver.anod_instance, 'test', False):
```
Comment (enzbang) [inline e3/electrolyt/run.py]: same here

File: e3/electrolyt/run.py [hunk 21]
```diff
@@ -29,6 +34,97 @@ def run(self):
...
                     getattr(self, method_name)()
 
+    def do_build(self):
+        """Run the build primitive after setting up the sandbox."""
+        self.data.data.sandbox = self.sandbox
+        anod_driver = AnodDriver(anod_instance=self.data.data, store=self.store)
+        anod_driver.activate()
+        anod_driver.anod_instance.build_space.create(quiet=True)
+        if not getattr(anod_driver.anod_instance, 'build', False):
+            logger.error('primitive build not implemented')
```
Comment (enzbang) [inline e3/electrolyt/run.py]: All error message should include enough info to debug the issue. Here you need to at least include the name of the spec.

File: e3/electrolyt/run.py [hunk 22]
```diff
@@ -29,6 +34,97 @@ def run(self):
...
+        self.status = STATUS.success
+
+    def do_checkout(self):
+        """Get sources from vcs to sandbox vcs_dir."""
+        repo_name = self.data.data[0]
+        repo_url = self.data.data[1]['url']
+        repo_revision = self.data.data[1]['revision']
+        repo_vcs = self.data.data[1]['vcs']
+        if repo_vcs != 'git':
+            logger.error('%s vcs type not supported' % repo_vcs)
```
Comment (enzbang) [inline e3/electrolyt/run.py]: replace % by ,

File: e3/electrolyt/run.py [hunk 23]
```diff
@@ -29,6 +34,97 @@ def run(self):
...
+        """Prepare src from vcs to cache using sourcebuilders."""
+        source_name = self.data.source_name
+        tmp_cache_dir = os.path.join(self.sandbox.tmp_dir, 'cache')
+        src = self.sandbox.vcs_dir
+        for src_builder in self.data.spec.source_pkg_build:
+            if src_builder.name == source_name:
+                repo_dict = {}
+                src_dir = os.path.join(src, src_builder.checkout[0])
+                dest_dir = os.path.join(tmp_cache_dir, source_name)
+                repo_dict[source_name] = {"working_dir": src_dir}
```
Comment (enzbang) [inline e3/electrolyt/run.py]: you're missing the repo state, if you want to do that in another commit please add:

  ??? missing repository state

and add an issue

File: e3/electrolyt/run.py (hunk 3 above)
Comment (enzbang) [inline e3/electrolyt/run.py]: typo?

File: tests/tests_e3/anod/test_sandbox.py [hunk 24]
```diff
@@ -66,6 +75,33 @@ def create_prolog(prolog_dir):
...
+def e3fake_git_dir(git, tmpdir):
+    """Create a fake e3 git repo for the test suite."""
+    del git  # skip or fail when git is not installed
+
+    # Create a e3-fake git repository
+    e3_fake_git_dir = str(tmpdir.mkdir('e3-fake-git-repo'))
+
+    try:
+        e3_fake_dir = os.path.join(os.path.dirname(__file__), 'e3-fake')
+        for source_file in os.listdir(e3_fake_dir):
```
Comment (enzbang) [inline tests/tests_e3/anod/test_sandbox.py]: can't you use e3.fs.cp or e3.fs.sync_tree?

File: tests/tests_e3/anod/action/main_test.py (hunk 15 above)
Comment (piratos) [inline tests/tests_e3/anod/action/main_test.py]: As it is in 
` return [Anod.SourceBuilder(
            name='e3-core-src',
            fullname=lambda x: 'e3-core-src.tgz',
            prepare_src=prepare_src,
            checkout=['e3-core-github'])]`
?

File: tests/tests_e3/anod/context_test.py (hunk 16 above)
Comment (piratos) [inline tests/tests_e3/anod/context_test.py]: debug leftover, removed

File: tests/tests_e3/anod/test_sandbox.py (hunk 24 above)
Comment (piratos) [inline tests/tests_e3/anod/test_sandbox.py]: Can I use part of the platform to test the other part ?

Comment (enzbang) [inline tests/tests_e3/anod/test_sandbox.py]: yes you can.

File: e3/anod/sandbox/action.py (hunk 10 above)
Comment (piratos) [inline e3/anod/sandbox/action.py]: thought the arg may pass other unexpected values, removed

File: e3/anod/sandbox/action.py (hunk 9 above)
Comment (piratos) [inline e3/anod/sandbox/action.py]: Fixed, I still add a debug line to know which resolver is finally used

File: e3/electrolyt/run.py (hunk 11 above)
Comment (piratos) [inline e3/electrolyt/run.py]: Done

File: e3/electrolyt/run.py (hunk 18 above)
Comment (piratos) [inline e3/electrolyt/run.py]: Added meaningful attributes names in place of "data", Done

File: e3/electrolyt/run.py (hunk 19 above)
Comment (piratos) [inline e3/electrolyt/run.py]: Done

File: e3/electrolyt/run.py (hunk 21 above)
Comment (piratos) [inline e3/electrolyt/run.py]: Done

File: e3/electrolyt/run.py (hunk 20 above)
Comment (piratos) [inline e3/electrolyt/run.py]: Done

File: e3/electrolyt/run.py (hunk 22 above)
Comment (piratos) [inline e3/electrolyt/run.py]: Done

File: e3/electrolyt/run.py (hunk 23 above)
Comment (piratos) [inline e3/electrolyt/run.py]: Done

File: e3/anod/action/__init__.py (hunk 12 above)
Comment (piratos) [inline e3/anod/action/__init__.py]: This is for the parent class which we don't know what kind of data it receives

File: e3/anod/status.py (hunk 13 above)
Comment (piratos) [inline e3/anod/status.py]: Done

File: tests/tests_e3/anod/action/main_test.py (hunk 14 above)
Comment (piratos) [inline tests/tests_e3/anod/action/main_test.py]: Done

File: e3/anod/context.py (hunk 17 above)
Comment (piratos) [inline e3/anod/context.py]: changed to 'anod_instance'

File: tests/tests_e3/anod/test_sandbox.py (hunk 24 above)
Comment (piratos) [inline tests/tests_e3/anod/test_sandbox.py]: Done
